REGRESSION (295715@main): [iOS] Safari extension popup windows may open scrolled down, website may flicker during scrolling

pxlcoder opened the pull request:
#### acd5bf8f6c4985b12b7d577f199f4dea5a50f3f1
<pre>
REGRESSION (295715@main): [iOS] Safari extension popup windows may open scrolled down, website may flicker during scrolling
<a href="https://bugs.webkit.org/show_bug.cgi?id=296056">https://bugs.webkit.org/show_bug.cgi?id=296056</a>
<a href="https://rdar.apple.com/155965298">rdar://155965298</a>

Reviewed by Abrar Rahman Protyasha.

295715@main inadvertently moved the zoom scale modification before computing the
scale ratio used to pin the scroll position to the top edge of the content. This
did not break PDFs since the scale ratio is computed using the ratio of the new
and old content sizes. However, the change effectively reverted 258521@main as
the scale ratio for all other web content would be 1, resulting in no content
offset adjustment, which is necessary to pin the scroll position. Instead, UIKit&apos;s
default behavior of pinning to the center occurs, resulting in unexpected jumps.

In the case of Safari extensions, the web view&apos;s width is increased during popover
presentation. This results in page scale changes for content with a fixed layout
width, and the scroll position may unexpectedly jump. Similarly, other sites, like
GitHub&apos;s Checks page, may flicker, as a wide-fixed element is added and removed
during scrolling. In that instance, page scale changes due to WebKit&apos;s shrink-to-fit
logic. This can result in an infinite loop where an element is added/removed based
on scroll position, changing the page scale, which then changes the scroll position.

Fix by moving the zoom scale modification back to where is originally was: after
computing the scale ratio, but before adjusting the content offset.

An API test is used to mimic the Safari extensions breakage, by simplying changing
the frame of the web view. This issue was not caught by the test added in 294990@main,
since that test only *decreased* the size. Additionally, 258521@main did not come
with any tests of its own. A layout test is used to mimic the flickering on GitHub.
Consequently, there now exists test coverage for scroll position pinning due to
page scale changes driven by frame changes, and WebKit&apos;s shrink-to-fit logic.

Thanks to Jessica Cheung for assisting in the investigation of this issue.

* LayoutTests/fast/scrolling/ios/pin-scroll-position-on-scale-change-should-not-flicker-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/pin-scroll-position-on-scale-change-should-not-flicker.html: Added.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AnimatedResize.mm:
(TEST(AnimatedResize, PinScrollPositionRelativeToTopEdgeOnPageScaleChangeAfterIncreasedSize)):

Canonical link: <a href="https://commits.webkit.org/299762@main">https://commits.webkit.org/299762@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c674f2e9c247797e8818c7cded460183889a491

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39679 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30330 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126316 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72059 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/aad24547-6c42-42dc-9410-a19e1698208a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121862 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48256 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91117 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60429 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b5bcfb29-be9f-4042-a133-3ec28b35fd01) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122938 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32248 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107590 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71672 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6a65365b-b5d8-4c5e-be95-effb016b4c1d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31279 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25695 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69952 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101723 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25883 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129236 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46906 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35568 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99736 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47272 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99581 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45036 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23054 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43527 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19087 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46768 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52474 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46234 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49583 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47920 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/AnimatedResize.mm:
```diff
@@ -594,6 +595,25 @@ - (void)_webView:(WKWebView *)webView didChangeSafeAreaShouldAffectObscuredInset
+    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 320, 500)]);
+    RetainPtr scrollView = [webView scrollView];
+
+    constexpr unsigned layoutWidth = 375;
```
Given this is static data
```suggestion
    static constexpr unsigned layoutWidth = 375;
```

aprotyas replied:
did you mean to re-use this for the frame width later too?

pxlcoder replied:
Done!

aprotyas reviewed (approved):
Typo in the commit message:
> by simplying changing

pxlcoder committed:
REGRESSION (295715@main): [iOS] Safari extension popup windows may open scrolled down, website may flicker during scrolling
https://bugs.webkit.org/show_bug.cgi?id=296056
rdar://155965298

Reviewed by Abrar Rahman Protyasha.

295715@main inadvertently moved the zoom scale modification before computing the
scale ratio used to pin the scroll position to the top edge of the content. This
did not break PDFs since the scale ratio is computed using the ratio of the new
and old content sizes. However, the change effectively reverted 258521@main as
the scale ratio for all other web content would be 1, resulting in no content
offset adjustment, which is necessary to pin the scroll position. Instead, UIKit's
default behavior of pinning to the center occurs, resulting in unexpected jumps.

In the case of Safari extensions, the web view's width is increased during popover
presentation. This results in page scale changes for content with a fixed layout
width, and the scroll position may unexpectedly jump. Similarly, other sites, like
GitHub's Checks page, may flicker, as a wide-fixed element is added and removed
during scrolling. In that instance, page scale changes due to WebKit's shrink-to-fit
logic. This can result in an infinite loop where an element is added/removed based
on scroll position, changing the page scale, which then changes the scroll position.

Fix by moving the zoom scale modification back to where is originally was: after
computing the scale ratio, but before adjusting the content offset.

An API test is used to mimic the Safari extensions breakage, by simplying changing
the frame of the web view. This issue was not caught by the test added in 294990@main,
since that test only *decreased* the size. Additionally, 258521@main did not come
with any tests of its own. A layout test is used to mimic the flickering on GitHub.
Consequently, there now exists test coverage for scroll position pinning due to
page scale changes driven by frame changes, and WebKit's shrink-to-fit logic.

Thanks to Jessica Cheung for assisting in the investigation of this issue.

* LayoutTests/fast/scrolling/ios/pin-scroll-position-on-scale-change-should-not-flicker-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/pin-scroll-position-on-scale-change-should-not-flicker.html: Added.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AnimatedResize.mm:
(TEST(AnimatedResize, PinScrollPositionRelativeToTopEdgeOnPageScaleChangeAfterIncreasedSize)):

Canonical link: https://commits.webkit.org/299762@main
